sysex reveng. on A4

yatli opened the pull request:
this is a work-in-progress.
uploading to share initial findings...

Commit:
begin sysex reveng. on A4

yatli commented:
uh oh I've sent 0x7C, 0x7D.

Measuring CV out now...

yatli commented:
CV out looks good, drives ms20 well.. doesn’t look like it’s a problem after the “calibration”

jmamma commented:
These are the references I know of:

https://github.com/bsp2/libanalogrytm

[mm1_22_sysex_0_6.pdf](https://github.com/jmamma/MIDICtrl20_MegaCommand/files/3007499/mm1_22_sysex_0_6.pdf)

[md1_53-sysex-0_8.pdf](https://github.com/jmamma/MIDICtrl20_MegaCommand/files/3007511/md1_53-sysex-0_8.pdf)

https://gitlab.com/dhuck/digitoner/blob/master/Digitoner.ipynb

It would be great if the encoding / decoding functionality could be made consistent across devices. This would either mean using Wesen's original implementation, or creating a new implementation and back porting to the MD + MNM

yatli commented:
MD command 0x40 is amazing.. I’m wondering if/how we could find that on other models..

Commit:
analyzer wip

Commit:
reveng: data collector ready. records req-rsp in records.json for later analytics

yatli commented:
enough banging.. collector's up. time to do some real job

I'm thinking about bringing in some formalism. For example with theorem provers, I can raise theories about the protocols, and the prover will try to prove me wrong, given the constraints and a database of samples.

yatli commented:
in case the prover is satisfied, we will have a working model of the protocol. the model may even contain parameters that are symbolically deduced from samples rather than given by hand..

Commit:
a4 reveng: recall display

yatli commented:
progress:

got a nice little protocol monitor;
did a quick and dirty session and found that trigger bits and note pitch info are recorded separately.

Trigger data is stored at the beginning:
![image](https://user-images.githubusercontent.com/20684720/55250072-6d0b5580-5288-11e9-827d-4a32a41595c0.png)

-- okay, one bit effective, among a few unknown bits :p

And notes later, at 0x9A:
![image](https://user-images.githubusercontent.com/20684720/55250339-eefb7e80-5288-11e9-9aa0-089aac4357b8.png)

Those `7F`s are fillers for inactive triggers.
`3C` is note `C4`
`48` is note `C5` -- standard midi

Commit:
uploading samples

jmamma commented:
Hi Yatao,

Can you make note of the size of A4 SYSEX dumps somewhere.

Thanks,

- Justin

Commit:
A4 data structures wip

yatli commented:
@jmamma the response sizes are attached to `A4Cmds.fs`.

Commit:
a4 reveng wip

yatli commented:
pattern & sound formats are partially decoded. see `Hypothesis.fs`

formal methods are overkill -- nothing fancy going on here.

jmamma commented on proto_analyzer/Hypothesis.fs:
```diff
@@ -0,0 +1,45 @@
+// TODO for both trigger and note pitch data, payloads are grouped in 7-byte blocks, separated by flipping 2a 55 2a 55s (acting as sync bytes?)
+//      00 in place of 2a/55 means no data in the block
+//      so, the raw bytes:  01 40 09 55 41 01 44 09 45 01 48 2a 09 49 01 4c 09 4d 01 55 50 09 51 01 54 09 55 2a
+//      are really:         01 40 09    41 01 44 09 45 01 48    09 49 01 4c 09 4d 01    50 09 51 01 54 09 55
```
This sounds like elektron's 7bit encoding method. To pack an array of bytes in to an array of 7 bit values, they take 7 bytes at a time. Each byte has the MSB removed, this MSB is stored in an extra byte for each of the 7 values.

So your 7bit data stream looks like:
[ MSBs ]  [ 7bitValue ]  [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]  (repeat)

"How the 7 bit encoding is generated
As a SysEx message only allows 7 bit data, 8 bit data blocks are converted as follows:
The first data (byte) contains the MSB of the following, up to, 7 data, data[0..6]. Bit 6 is the MSB bit of data[0] and bit 0 is the MSB of data[6]. 7 bytes encodes to 8 data. 8 bytes encodes to 10 data."

jmamma replied:
I should add. Please refer to the mnm pdf i uploaded explaining packing and encoding.

yatli replied:
thanks, that's very interesting. I'll read up the mnm manual to follow up.
the MSBs probably aren't used here though, as the truth table already covers all the information for triggers.
also, please note:
```
0x55 = 0101 0101
0x2a = 0010 1010
```
So they add up to 0x7F (and do not chance however I arm the triggers) -- more like a synchronization pattern -- in case some bytes are dropped, it still recovers "on the beat".
My reasoning goes the same with the 01-09 pattern.
Another explaination would be, that all the effective bytes (0x40, 0x41 etc) have the MSB toggled, and the 01-09 patterns do not. If so, because each block contains 7 bytes, the pattern naturally shuffles like `1010101   0101010  1010101  0101010 ... `.

it surprised me that the 7bit and 8bit COM protocols are compatible, so the monitor program works without even touching the width and stop bit settings. :p

jmamma commented:
> pattern & sound formats are partially decoded. see `Hypothesis.fs`
> 
> formal methods are overkill -- nothing fancy going on here.

great work so far.

Commit:
calibrate a4 types. update sound patch captures to identify

yatli commented:
it's a bit boring to make 100 patches to identify the params but there we have it :)
patch 0 is the reference setting, and each of the following have one param tweaked.
the order is noted in `A4Types.fs`, namely `type Sound = { ... }` record type.

the patches are extracted and added to `samples.7z`.

there's one part at the end of each patch that I don't understand yet (offset 0x19a-0x19b):

![image](https://user-images.githubusercontent.com/20684720/55372971-30fd1c80-5536-11e9-8759-3d205f046d88.png)

these two bytes keep changing on all the patches. could be a checksum of some kind. does not look like a shuffling scheme though, because the values are so concentrated around 0x3d00-0x3f00, and more tweaks in the patch = higher checksum.

might be just a direct sum over the bytes 0.0

jmamma commented:
See the MonoMachine pdf for how checksums work:

```
How the checksums are calculated
The checksum is generated from the data marked in blue in the tables. It is the lower 14 bits of the sum of all the indicated data.

[0 ... 127] Checksum (bit 7...13)
[0 ... 127] Checksum (bit 0...6)
[0 ... 127] Message length (bit 7...13)
[0 ... 127] Message length (bit 0...6)
0xF7 End of SysEx
```

yatli commented:
great! with that piece of information installed, I can start writing a c++ prototype for sound patches very soon.

Commit:
A4Messages: sound type wip

yatli commented:
update:

working on the prototype now; please see `A4Messages{.h,.cpp}`

also, the analyzer is giving me positive results. please see `sounds.inference.txt`

yatli commented:
have to say, this is a bit convoluted:

```c++
  decoder.skip(1); // skip sound header 0x05
  decoder.get(tags);
  decoder.get(name);
  decoder.get(osc[0].tuning);
  decoder.get(osc[0].fine);
  decoder.get(osc[1].tuning);
  decoder.get(osc[1].fine);
  decoder.get(osc[0].detune);
  decoder.get(osc[1].detune);
  decoder.get(osc[0].keytrack);
  decoder.get(osc[1].keytrack);
```

is it ok to use the decoder like this?
from what you're doing in the other PR, I guess batch reads are more effective.

jmamma commented:
> update:
> 
> working on the prototype now; please see `A4Messages{.h,.cpp}`
> 
> also, the analyzer is giving me positive results. please see `sounds.inference.txt`

Just had quick glance. Amazing. looking good so far!

jmamma commented:
> have to say, this is a bit convoluted:
> 
> ```c++
>   decoder.skip(1); // skip sound header 0x05
>   decoder.get(tags);
>   decoder.get(name);
>   decoder.get(osc[0].tuning);
>   decoder.get(osc[0].fine);
>   decoder.get(osc[1].tuning);
>   decoder.get(osc[1].fine);
>   decoder.get(osc[0].detune);
>   decoder.get(osc[1].detune);
>   decoder.get(osc[0].keytrack);
>   decoder.get(osc[1].keytrack);
> ```
> 
> is it ok to use the decoder like this?
> from what you're doing in the other PR, I guess batch reads are more effective.

The decoder code is all of Wesen's original work. Batch reads should be okay and will be more efficient, and reduce program size.

A long, long time ago I fixed a bug MDPattern.cpp relating to batch reads.
 Looking at the code now, I can't actually work out why it failed, but the pattern trig masks would end up incomplete.

Original code (https://github.com/wesen/mididuino/blob/master/hardware/libraries/MD/MDPattern.cpp):
```
  decoder.start7Bit();
  decoder.get32(lockPatterns, 16);

  decoder.start7Bit();
  decoder.get32(&accentPattern, 4);
  /*
    accentPattern = decoder.gget32();
    slidePattern  = decoder.gget32();
    swingPattern  = decoder.gget32();
    swingAmount   = decoder.gget32();
  */
  decoder.stop7Bit();
```

Patched code (MDPattern.cpp):

```
111     decoder.start7Bit();
112     decoder.get32(lockPatterns, 16);
113     
114     decoder.start7Bit();
115     
116     decoder.get32(&accentPattern);
117 
118     decoder.get32(&slidePattern);
119     decoder.get32(&swingPattern);
120 
121     decoder.get32(&swingAmount);
122 
123     /*
124      accentPattern = decoder.gget32();
125      slidePattern  = decoder.gget32();
126      swingPattern  = decoder.gget32();
127      swingAmount   = decoder.gget32();
128      */
129     decoder.stop7Bit();
```

yatli commented:
Hmm, I suspect it's call start7Bit() twice? The state may be lagging behind and the decoder ate some bytes.

Commit:
cleaning up sound patch inference results

yatli commented:
I've cleaned up the inference results.
there were noisy bits possibly due to encoder sensitivity (reset to 0, then bounced to 1)

I'm going to figure out how to efficiently decode them.

Commit:
implementation unblocked

Commit:
finish analyzing

Commit:
a4 sound codec complete.

yatli commented:
@jmamma I've rolled the parameters into a big struct. that should make the codec process more efficient. I think the codec is up for some testing!

jmamma commented:
Amazing.

> @jmamma I've rolled the parameters into a big struct. that should make the codec process more efficient. I think the codec is up for some testing!

I think this will be okay for A4Sound. I was thinking yesterday that A4Pattern at 14KB is going to be too big to keep around as global object. We may need to break it down in to loadable sections.

Commit:
a4 sound codec in kit

yatli commented:
> I was thinking yesterday that A4Pattern at 14KB is going to be too big to keep around as global object. We may need to break it down in to loadable sections.
they're not packing things as tight as possible. we can squeeze ~40% out without compression.
with compression I'm not sure. but maybe not worth it.

can we at least fit it into the sysex recv buffer?

yatli commented:
another approach, if we can afford decoding just-in-time while sysex is receiving, then we decode the tracks one by one reducing the sysex size to 25%

jmamma commented:
> they're not packing things as tight as possible. we can squeeze ~40% out without compression.
> with compression I'm not sure. but maybe not worth it.
> 
> can we at least fit it into the sysex recv buffer?

The pull request (sysex refactor) that I'm working on is to move to sysex buffers to the HIGH memory bank. This has two benefits, it will allow us to process much large sysex dumps, and secondly free's up more main memory for timing critical code, like the sequencer.

yatli commented:
ah ok, no need to do JIT decoding. leave more space for the sequencer is the right way to go.

yatli commented:
While you're on the sysex infrastructure, I will work out A4kit/A4pattern codec. :)
Let's see how much space we can squeeze from it.

jmamma commented:
I just got the ArduinoIDE to compile with new sysex code and we have 
17170 Bytes of RAM free.

Not bad :) .

jmamma committed:
Merge branch 'dev' into dev

jmamma commented:
Finally had time to look through the code. Great work here.

jmamma committed:
Simplify ISR code, split isr_uart and isr_uart2 code

jmamma committed:
simplify transmit isr

jmamma committed:
eliminate division from div192th calc

jmamma committed:
Remove suspect again statements

jmamma committed:
Don't re-enable ISR in MIDI callbacks

jmamma committed:
Don't call, CC noteon/off callbacks when sending MIDI data

This seems to be a big source of slowdown in the isr.

jmamma committed:
MidiClock callbacks can now be immediately in uart_isr on clock pulse

jmamma committed:
Don't calculate tempo in isr. calculate on demand if out of date

jmamma committed:
re-enable isr in callbacks

jmamma committed:
some changes to ensure atomic access/modification of memory

jmamma committed:
inline some isr more code

jmamma committed:
disable sanity checking for MIDI messages

jmamma committed:
send debug message on bad kit length

jmamma committed:
Let's re-enable kit saving when sequencer is running and patch the kit original values with what they should be

Commit:
Merge branch 'dev' of github.com:jmamma/MIDICtrl20_MegaCommand into dev

jmamma committed:
Dont debug div16th_counter in GridTask

jmamma committed:
If sequencer is running, dont receive pattern/kit on 16th.

jmamma committed:
Don't fetch pattern if sequencer running. Don't allow sequencer merge if sequencer is running

jmamma committed:
Chain mode is always active

jmamma committed:
Strip out deprecated write mode code

jmamma committed:
regression: chain rnd mode was hidden

jmamma committed:
Don't use delay for TurboLight, try to follow the spec by sending 16 bytes

jmamma committed:
Changes to save and write modes.

No longer send pattern to MD.
MD Sequencer Merge is now selectable from the save page and only
available when the sequencer is not running.
Removed convoluted write/store in specific slots behaviours

jmamma committed:
Bug Fix: Current track length would be accidenlty set to last track lenght on StepPage enter

jmamma committed:
Fast timer halve execution frequencer from 10000Hz to 5000Hz, giving us 0.2ms accuracy

jmamma committed:
Fixed CuePage. Works with internal sequencer

jmamma committed:
Cue Page is now a generalised Audio Routing page

jmamma committed:
Improvements to route page interaction with sequencer

jmamma committed:
Fix crash when switching between pages

Commit:
update .ccls

Commit:
optimize bit arithmetics

Commit:
optimize _BV32

Commit:
remove PROGMEM on mask LUTs

jmamma committed:
Fix issue with 8bit manipulation

jmamma committed:
Add brackets to fix random behaviour, and compiler warnings

Commit:
wip

Commit:
cleanup.

jmamma committed:
optimise conditional trigs, by removing division.

Commit:
xor_rand

jmamma committed:
Make hold time consistant across pages and fix overflow

jmamma committed:
Remove mod operator from sequencer code

jmamma committed:
use fast get_random_byte() random function

jmamma committed:
Optimise conditional trig code for ext tracks

jmamma committed:
loop jaming

jmamma committed:
optimise multiplication

jmamma committed:
remove condition + timing variables and remove divison

jmamma committed:
remove duplicate trig

jmamma committed:
bug fix

jmamma committed:
forgot to checking main.cpp

jmamma committed:
Forgot to update params if sequencer is running on save.

If not, kit values might be saved to current parameter lock

jmamma committed:
Ugly optimization by pushing sequencer code within the seq function.

jmamma commented:
Thought I had already merged this one..

jmamma committed:
Change note buffer data structure to bitmask with btree search.

Note buffer array data structure is used to store notes
that the sequencer has sent to device. Therefor when the sequencer
is stopped, note_off events for these notes can be sent.

The original implentation was an array which involved loops within
the seq isr. The new method avoids isr loops by taking advantage of bit masks.

To scan for active notes we use a btree search method.

jmamma committed:
bug fix: offset not calculated correctly

jmamma committed:
Revise binary tree. Checking branches has no computation advantage until
you hit uint8_t.

jmamma committed:
Auto chain was broken for A4 tracks

jmamma committed:
manual mode would still chain automatically in some circumstances

jmamma committed:
MidiClock continue was broken causing tempo to be miscalculated

jmamma committed:
remove duplicate calculation of transitions on start

jmamma committed:
offset calculation was wrong

jmamma committed:
I hate pointers, sometimes

jmamma committed:
remove note on debug

jmamma committed:
fix char 005 for SeqExtStep page.

jmamma committed:
Revert "Ugly optimization by pushing sequencer code within the seq function."

We can do this optimisaiton at a later stage, if we determine the
performance benefits outweigh the code clutter.

This reverts commit 8b2825443f8b9bd90fd18c66d0db60cd81640a8b.

jmamma committed:
A4 latency calculation was not correct causing unmute to fail on chain

jmamma committed:
Colours could be inverted when leaving GridPage in some circumstances

jmamma committed:
fix display flicker on SeqExtStep page

jmamma commented:
```/Applications/Arduino.app/Contents/Java/hardware/valence/avr/cores/megacommand/A4/A4Messages.hh:8:21: fatal error: algorithm: No such file or directory
 #include <algorithm>
                     ^
````

Getting compile error

jmamma commented:
Do we need to include a c++ STL?

Were you using something like ?

https://github.com/maniacbug/StandardCplusplus

jmamma commented:
```
   a4sfloat_t(uint8_t dat[2]) { std::copy(dat, dat + 2, data); }```

I think we should avoid using the STL, due to increase in program size.

For example, the implementation above could be done using memcpy, which is faster than a for routine with pointers.

yatli commented:
yeah I was in a STL mood back then but it's obviously not ideal to link against libc++!

jmamma committed:
Remove reference to C++ STL

jmamma commented:
I removed the reference to C++ STL.

I realised that the A4Sound class has changed. Which could possibly conflict with the byte layout of the old A4Sound, leading to incompatibility ?

Also A4Sound::fromSysex(MidiClass *midi) { } needs to be updated.

jmamma committed:
Fixes for fromSysex

jmamma committed:
fixed compile errors

jmamma commented:
This now compiles , but saving and sending sounds from MCL does not appear to be working correctly.
Need to check fromSysex()

jmamma committed:
Update A4Messages.cpp

Fix typo

jmamma commented:
New A4 firmware. 1.40 released.

Need to check compatibility.

yatli commented:
Wow, conflicts.
Needs a good deep dive into this thread after the MD side.

Commit:
Merge branch 'upstream_dev' into dev

Commit:
cleanup.

yatli commented:
@jmamma this branch is up-to-date with `dev`. I can resume the breakpoint now.

Commit:
fix build error

Commit:
A4Sound: fix a4sound_t layout errors. fix A4 sound footer error. fix transmission logic.

yatli commented:
good news. A4 recognizes the transmission as sound patch!

now just have to fix the incompatibility between old blob format and the new struct.

yatli commented on avr/cores/megacommand/A4/A4Messages.cpp:
```diff
@@ -58,7 +58,7 @@ uint16_t A4Global::toSysex(ElektronDataToSysexEncoder &encoder) {
 static constexpr uint8_t a4soundx_prologue[8] = { 0x00, 0x20, 0x3c, 0x06, 0x00, 0x59, 0x01, 0x01 };
 static constexpr uint8_t a4sound_header[8] = { 0x78, 0x3e, 0x6f, 0x3a, 0x3a, 0x00, 0x00, 0x00 };
-static constexpr uint8_t a4sound_footer[4] =  { 0x3a, 0xe3, 0x7a, 0x4e };
+static constexpr uint8_t a4sound_footer[4] =  { 0x3a, 0x3e, 0x7a, 0x4e };
```
#FAIL

yatli commented:
@jmamma we should decide whether to keep the compatibility with the current A4 patches.

Commit:
A4sound: attempt to support sysex dump format

yatli commented:
Tested, mixed result. Some patches can be loaded properly, others result in a "blank" patch, everything stuck at leftmost position (-64, 0, off, etc)

yatli commented:
I've written a simple predicate to test whether a patch is legacy blob or structured.
See A4Sound::is_legacy_format()

jmamma commented:
> @jmamma we should decide whether to keep the compatibility with the current A4 patches.

What's involved to do this? Maintain two classes?

I upgraded A4 firmware a few weeks ago, hadn't noticed sound dump incompatibility. But my testing was not thorough.

yatli commented:
The sysex dumps from A4 haven't changed. I've replaced the blob payload in `A4Sound` with structured parameters, and that strips out the 7bit encoding.

yatli commented:
The hack:

```C++

  // === Begin new a4 sound patch layout
  uint8_t          tags[4];      // 32 tags
  char             name[16];     // null-terminated
  // sizeof(a4sound_t) is 318, which encodes 363 bytes in 7-bit enc, + 42 bytes header metadata + 8B prologue + 2B SYSEX frame = 415B
  a4sound_t        sound;

  // === End new a4 sound patch layout, 338B

  // old a4 sound patch layout: uint8_t[415 - 10 - 2 - 4 - 1], 398 bytes
  // -10 : prologue and 0x78 0x3E, part of a4 sound header.
  // -2: sysex frame
  // -4: len & checksum
  // -1: origposition
  uint8_t          sound_pad[60];

  bool is_legacy_format();
  void fromLegacySysexDump();

bool A4Sound::is_legacy_format() {
  uint8_t* payload = this->tags;
  return !memcmp(payload, a4sound_header + 2, 6);
}

void A4Sound::fromLegacySysexDump() {
  toggleLed2();
  // legacy payload len = 398
  uint8_t* payload = this->tags;
  // skip the partial header 
  payload += 6;
  // getting data from itself. always a few steps behind.
  ElektronSysexDecoder decoder(DATA_ENCODER_INIT(payload, 392));
  fromSysex_impl(decoder);
}
```

jmamma commented:
Ah yeah. Okay. 

This is probably a more broader issue. Do we support legacy project formats ?

For instance, there are changes id like to make the MDTrack structure, but i've been holding off because it will break all my projects.

Commit:
proto_analyzer: support MonoMachine

yatli commented:
Is the project format versioned? If so we can perform one-way upgrade on it without maintaining different versions of actual working classes.

jmamma commented:
I'm thinking we stick with legacy sound dump until version 3.0

Major version change should == project spec change.

jmamma commented:
> Is the project format versioned? If so we can perform one-way upgrade on it without maintaining different versions of actual working classes.

Yes it is.

Commit:
MonoMachine id

Commit:
add serial monitor

jmamma reviewed (changes requested):
You mean DEVICE_MNM ?

jmamma commented on avr/cores/megacommand/Midi/MidiID.hh:
```diff
@@ -7,6 +7,7 @@
 #define DEVICE_NULL 0xFF
 #define DEVICE_MIDI 0xFE
 #define DEVICE_MD 0x02
+#define DEVICE_MD 0x03
```
DEVICE_MNM ?

jmamma replied:
Also FYI MidiID does not work on the new machines last time I checked.

yatli replied:
If so we can use a MNM-specific sysex query as a probe.
Newer machine=digitakt etc?

Commit:
Merge branch 'ext_conv' into dev

yatli commented:
found a better home for this branch.

Commit:
update

Commit:
Merge branch 'ext_conv' into dev

Commit:
minor

Commit:
Merge branch 'ext_conv' into dev

Commit:
adapt to new convert method prototype
